nmdc_schema/migrators/migrator_from_X_to_PR19_and_PR70.py

```diff
@@ -0,0 +1,89 @@
+        self.adapter.process_each_document(collection_name="omics_processing_set", pipeline=[self.transform_instrument_slot]) 
+
+    def fill_instrument_set(self):
+        r""""
+        Fills the instrument_set collection from a YAML file of curated instrument instances
+        
+        >>> from nmdc_schema.migrators.adapters.dictionary_adapter import DictionaryAdapter
+        >>>
+        >>> database = {"instrument_set": []}  # seeds the database
+        >>> adapter = DictionaryAdapter(database=database)
+        >>> m = Migrator(adapter=adapter)
+        >>> m.fill_instrument_set()
+        >>> result = adapter.get_document_having_value_in_field('instrument_set', 'name', value='12T FT-ICR MS')
+        >>> result['name']
+        '12T FT-ICR MS'
+        >>> result['vendor']
+        'bruker'
+        >>> result['model']
+        'solarix_12T'
+        >>> result['type']
+        'nmdc:Instrument'
+
+        """
+
+        instruments = load_yaml_asset("migrator_from_X_to_PR19_and_PR70/instrument_set.yaml")
+        for instrument in instruments:
+            self.adapter.insert_document("instrument_set", instrument)
+
+    def transform_instrument_slot(self, omics_doc: dict):
+        r"""
+        Changes the instrument_name slot to be instrument_used, and replaces the name string value with an identifier for 
+        the instrument
+
+        >>> from nmdc_schema.migrators.adapters.dictionary_adapter import DictionaryAdapter
+        >>>
+        >>> database = {'instrument_set': [{'id': 'nmdc:inst-456', 'name': '12T FT-ICR MS', 'model':'solarix_12T', 'vendor':'bruker', 'type':'nmdc:Instrument'}]} 
+        >>> adapter = DictionaryAdapter(database=database)
+        >>> m = Migrator(adapter=adapter)
+        >>> m.transform_instrument_slot({'id': 'nmdc:omcp-123', 'instrument_name': '12T_FTICR_B'})
+        {'id': 'nmdc:omcp-123', 'instrument_used': 'nmdc:inst-456'}
+        """
+
+        if "instrument_name" in omics_doc:
+            existing_instrument_name = omics_doc["instrument_name"]
+            
+            instruments = load_yaml_asset("migrator_from_X_to_PR19_and_PR70/instrument_set.yaml")
+            allowed_instrument_names = [instrument["name"] for instrument in instruments]
+            
+            matches = difflib.get_close_matches(existing_instrument_name, allowed_instrument_names, n=1, cutoff=0.25)
+
+            if matches:
+
+                    # Match instrument_name with with instrument set and get corresponding instrument id
+                    match = matches[0]
+                    self.logger.info(f"instrument_name in {omics_doc['id']} is {omics_doc['instrument_name']} and will be matched to instrument with name {match}")
+                    
+                    instrument = self.adapter.get_document_having_value_in_field(
+                        collection_name="instrument_set", field_name="name", value=match
+                    )
+
+                    instrument_id = instrument.get("id")
```

Comment:
I want to check my understanding of what is happening in this migrator. 

The `instrument_set` collection was created (via the `upgrade` method). Then, it was populated with the instruments listed in the YAML file (via the `fill_instrument_set` method). Finally, in this method, the migrator is using the YAML file again (not the database) to get the distinct instrument names and then using the database (not the YAML file) to get the `id` value of a specific instrument.

I assume the YAML file (instead of the database) was used as the source of the names because `self.adapter` doesn't provide a "clean" way to get all the names–it would require using the `self.adapter.process_each_document` method to gather all the names and store them in a local list. I'll keep this use case in mind for the next iteration of adapter development 🧑‍💻 .

Reply:
Yep, this is how it works. And yes, I guess I decided to use the yaml file because it was an easy way to get a list of all the instrument names. I supposed I could have used the `process_each_document` adapter on the `instrument_set` to append each instrument name to a list and use that as the acceptable instrument names.

Reply:
Thanks for confirming. I'm OK with the way it's currently implemented.